client/rpc-spec-v2/src/chain_head/chain_head_storage.rs

```diff
@@ -0,0 +1,189 @@
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + 'static,
+{
+	/// Fetch the value from storage.
+	fn query_storage_value(
+		&self,
+		hash: Block::Hash,
+		key: &StorageKey,
+		child_key: Option<&ChildInfo>,
+	) -> Option<QueryResult> {
+		let result = if let Some(child_key) = child_key {
+			self.client.child_storage(hash, child_key, key)
+		} else {
+			self.client.storage(hash, key)
+		};
+
+		result
+			.map(|opt| {
+				opt.map(|storage_data| {
+					QueryResult::Buffered(StorageResult::<String> {
+						key: format!("0x{:?}", HexDisplay::from(&key.0)),
+						result: StorageResultType::Value(format!(
+							"0x{:?}",
+							HexDisplay::from(&storage_data.0)
+						)),
+					})
+				})
+			})
+			.unwrap_or_else(|err| {
+				Some(QueryResult::Immediate(ChainHeadStorageEvent::<String>::Error(ErrorEvent {
+					error: err.to_string(),
+				})))
+			})
+	}
+
+	/// Fetch the hash of a value from storage.
+	fn query_storage_hash(
+		&self,
+		hash: Block::Hash,
+		key: &StorageKey,
+		child_key: Option<&ChildInfo>,
+	) -> Option<QueryResult> {
+		let result = if let Some(child_key) = child_key {
+			self.client.child_storage_hash(hash, child_key, key)
+		} else {
+			self.client.storage_hash(hash, key)
+		};
+
+		result
+			.map(|opt| {
+				opt.map(|storage_data| {
+					QueryResult::Buffered(StorageResult::<String> {
+						key: format!("0x{:?}", HexDisplay::from(&key.0)),
+						result: StorageResultType::Hash(format!("{:?}", storage_data)),
+					})
+				})
+			})
+			.unwrap_or_else(|err| {
+				Some(QueryResult::Immediate(ChainHeadStorageEvent::<String>::Error(ErrorEvent {
+					error: err.to_string(),
+				})))
+			})
+	}
+
+	/// Make the storage query.
+	fn query_storage(
+		&self,
+		hash: Block::Hash,
+		query: &StorageQuery<StorageKey>,
+		child_key: Option<&ChildInfo>,
+	) -> Option<QueryResult> {
+		if !is_key_queryable(&query.key.0) {
+			return None
+		}
+
+		match query.ty {
+			StorageQueryType::Value => self.query_storage_value(hash, &query.key, child_key),
+			StorageQueryType::Hash => self.query_storage_hash(hash, &query.key, child_key),
+			_ => None,
+		}
+	}
+
+	/// Generate the block events for the `chainHead_storage` method.
+	pub fn generate_events(
+		&self,
+		mut sink: SubscriptionSink,
+		hash: Block::Hash,
+		items: Vec<StorageQuery<StorageKey>>,
+		child_key: Option<ChildInfo>,
+	) {
+		if let Some(child_key) = child_key.as_ref() {
+			if !is_key_queryable(child_key.storage_key()) {
+				let _ = sink.send(&ChainHeadStorageEvent::<String>::Done);
+				return
+			}
+		}
+
+		let mut storage_results = Vec::with_capacity(items.len());
+		for item in items {
+			let Some(result) = self.query_storage(hash, &item, child_key.as_ref()) else {
+                continue
+            };
+
+			match result {
+				QueryResult::Buffered(storage_result) => storage_results.push(storage_result),
+				QueryResult::Immediate(event) => {
+					let _ = sink.send(&event);
```

Comment:
The spec says this about errors: "This can only be the first event generated by this subscription. No other event will be generated with this subscription.". As far as I see we continue to send events here even after some error was sent.

Reply:
Thanks for pointing this out! I opened https://github.com/paritytech/json-rpc-interface-spec/issues/63 in the spec to ask for some clarifications here, since we could have multiple items in the query we might still want to return the `Error` after processing and sending some results back to the user

Reply:
Oki doki, have made the error event here the first event and last event.
No more events would be produced now after we encounter an error here

I've opened https://github.com/paritytech/json-rpc-interface-spec/issues/63 for further discussion about intermediate error events.

In the meanwhile, to make some progress here we'll stick with today's version of the spec
Thanks!